src/ConstructiveSolidGeometry/VolumePrimitives/Cone.jl

```diff
@@ -1,207 +1,850 @@
+function _in(pt::CartesianPoint, c::PartialVaryingCylinder{T,ClosedPrimitive}) where {T} 
+    az = abs(pt.z) 
+    (az <= c.hZ || csg_isapprox_lin(az, c.hZ)) && begin
+        r = hypot(pt.x, pt.y) 
+        rz = radius_at_z(c.hZ, c.r[1], c.r[2], pt.z)
+        r <= rz || csg_isapprox_lin(r, rz)
+    end &&
+    _in_angular_interval_closed(atan(pt.y, pt.x), c.φ)
+end
+function _in(pt::CartesianPoint, c::PartialVaryingCylinder{T,OpenPrimitive}) where {T} 
+    abs(pt.z) < c.hZ &&
+    hypot(pt.x, pt.y) < radius_at_z(c.hZ, c.r[1], c.r[2], pt.z) &&
+    _in_angular_interval_open(atan(pt.y, pt.x), c.φ)
 end
 
-function Tube(rMin::R1, rMax::R2, height::H) where {R1<:Real, R2<:Real, H<:Real}
-    T = float(promote_type(R1,R2,H))
-    Cone(T, rMin == 0 ? T(rMax) : T(rMin)..T(rMax), nothing, T(height)/2)
+function surfaces(t::PartialVaryingCylinder{T}) where {T}
+    bot_center_pt = _transform_into_global_coordinate_system(CartesianPoint{T}(zero(T), zero(T), -t.hZ), t) 
+    top_center_pt = _transform_into_global_coordinate_system(CartesianPoint{T}(zero(T), zero(T), +t.hZ), t) 
+    mantle = ConeMantle{T,Tuple{T,T},Tuple{T,T},:inwards}((t.r[1], t.r[2]), t.φ, t.hZ, t.origin, t.rotation)
+    e_bot = PartialCircularArea{T}(r = t.r[1], φ = t.φ, origin = bot_center_pt, rotation = t.rotation)
+    e_top = PartialCircularArea{T}(r = t.r[2], φ = t.φ, origin = top_center_pt, rotation = -t.rotation * RotZ{T}(π))
+    poly_l = _transform_into_global_coordinate_system(Quadrangle{T}((
+        CartesianPoint(CylindricalPoint{T}(zero(T),  t.φ[1], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(zero(T),  t.φ[1], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}( t.r[2],  t.φ[1], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}( t.r[1],  t.φ[1], -t.hZ)) )), t)
+    poly_r = _transform_into_global_coordinate_system(Quadrangle{T}((
+        CartesianPoint(CylindricalPoint{T}(zero(T),  t.φ[2], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}( t.r[1],  t.φ[2], -t.hZ)),
+        CartesianPoint(CylindricalPoint{T}( t.r[2],  t.φ[2], +t.hZ)),
+        CartesianPoint(CylindricalPoint{T}(zero(T),  t.φ[2], +t.hZ)) )), t)
+    # normals of the surfaces show inside the volume primitives. 
+    e_top, e_bot, mantle, poly_l, poly_r
 end
 
-# for Tubes
-get_r_at_z(c::Cone{T, <:Union{T, AbstractInterval{T}}, <:Any, <:Any}, z::Real) where {T} = c.r
 
-# for Cones
-get_r_at_z(c::Cone{T, Tuple{T,T}, <:Any, <:Any}, z::Real) where {T} = _get_r_at_z(c.r[1], c.r[2], c.z, z)
 
-function get_r_at_z(c::Cone{T, Tuple{I,I}, <:Any, <:Any}, z::Real) where {T, I<:AbstractInterval{T}}
-    r1::T = _get_r_at_z(c.r[1].left, c.r[2].left, c.z, z)
-    r2::T = _get_r_at_z(c.r[1].right, c.r[2].right, c.z, z)
-    r1..r2
+####################################################################
+####################################################################
+
+#=
+  ___
+ /   \
+/_____\
+
+=#
+const VaryingTube{T,CO} = Cone{T,CO,Tuple{Tuple{T,T},Tuple{T,T}},Nothing} # Full in φ
```

Comment:
`Tube` is a `Cone` with a non-zero `rmin`?

Reply:
Yes, and `r_min` != `r_max` for top and bot.